core/src/main/scala/org/apache/spark/SparkContext.scala

```diff
@@ -91,6 +91,7 @@ class SparkContext(config: SparkConf) extends Logging {
   val startTime = System.currentTimeMillis()
 
   private[spark] val stopped: AtomicBoolean = new AtomicBoolean(false)
+  private[spark] val sessionListenerRegistered: AtomicBoolean = new AtomicBoolean(false)
```

Comment:
can we move this to `object SparkSession`?

Reply:
Not easily - that was my original plan, however a `SparkSession` object can be constructed with an instance of `SparkContext`. Meaning that it's really on the `SparkContext` to know whether a session has already attached a listener. The `SparkSession` instance can then decide whether or not to attach the listener, but the context will need to report whether the listener is attached first.

Reply:
Sorry if I missed anything. `object SparkSession` is a singleton and it can make sure we do something only once. Are you saying that we can register a session cleanup listener to `SparkContext` outside of `SparkSession`.

Reply:
Yeah so the challenge is that even though `SparkSession` is a singleton, I can create various spark session instances that are all "backed" by a different spark context instance. 
```    
val conf = new SparkConf()
      .setMaster("local")
      .setAppName("test-app-SPARK-31354-1")
    val context = new SparkContext(conf)

    val session1 = SparkSession.builder()
      .master("local")
      .sparkContext(context) // <- here
      .getOrCreate()
```
So, the state actually needs to moreso be on the context, since the context is the only one that knows whether it has already had a session listener attached. I agree that this feels really weird, and that's kind of why I was initially pursuing adding lifecycle methods to encapsulate the state in one place. Definitely open to other ways to resolve this though.

Reply:
I think we can assume there is only one `SparkContext` instance. See https://github.com/apache/spark/pull/23311

One problem is we create a `SparkContext`, stop it and create it again. Then it's not safe to rely on the flag in singleton `object SparkSession`

Reply:
If we really want to guide against the corner case, we can track the `SparkContext` instance id in `object SparkSession`. If `SparkContext` is re-created, we can reset the boolean flag.

Reply:
@cloud-fan - updated. how does this look?

Reply:
@cloud-fan - updated, how does this look?